### PR TITLE
my improvements for httpcore

### PR DESCRIPTION
**Pull request by Araq**

### CODE REVIEW

File: lib/pure/httpcore.nim [hunk 1]
```diff
@@ -99,23 +99,25 @@ const
...
-    pairs.add((pair.key.toLowerAscii(), @[pair.val]))
-  new result
-  result.table = newTable[string, seq[string]](pairs)
+    HeadersImpl(result)[pair.key.toLowerAscii()] = @[pair.val]
+
+proc `$`*(headers: HttpHeaders): string = $(HeadersImpl(headers))
 
-proc `$`*(headers: HttpHeaders): string =
-  return $headers.table
+proc isEmpty*(a: HttpHeaders): bool = HeadersImpl(a) == nil
```
Comment (dom96) [inline lib/pure/httpcore.nim]: This implies that `[]` on it will work, but will it?

File: lib/pure/httpcore.nim [hunk 2]
```diff
@@ -99,23 +99,25 @@ const
 
 const headerLimit* = 10_000
 
+template table*(x: HttpHeaders): TableRef[string, seq[string]] {.
+    deprecated: "use the other accessor procs instead".} =
+  TableRef[string, seq[string]](x)
```
Comment (dom96) [inline lib/pure/httpcore.nim]: This needs a doc comment to specify it's deprecated.

File: lib/pure/asynchttpserver.nim [hunk 3]
```diff
@@ -78,7 +78,7 @@ proc sendHeaders*(req: Request, headers: HttpHeaders): Future[void] =
   return req.client.send(msg)
 
 proc respond*(req: Request, code: HttpCode, content: string,
-              headers: HttpHeaders = nil): Future[void] =
+              headers = HttpHeaders(nil)): Future[void] =
```
Comment (dom96) [inline lib/pure/asynchttpserver.nim]: Maybe introduce a `EmptyHttpHeaders` in `httpcore`?

File: lib/pure/httpcore.nim [hunk 4]
```diff
@@ -172,18 +175,15 @@ proc contains*(values: HttpHeaderValues, value: string): bool =
...
 
 proc getOrDefault*(headers: HttpHeaders, key: string,
     default = @[""].HttpHeaderValues): HttpHeaderValues =
   ## Returns the values associated with the given ``key``. If there are no
   ## values associated with the key, then ``default`` is returned.
-  if headers.hasKey(key):
-    return headers[key]
-  else:
-    return default
+  result = HttpHeaderValues HeadersImpl(headers).getOrDefault(key, seq[string](default))
```
Comment (dom96) [inline lib/pure/httpcore.nim]: I really wish this had more parenthesis :)

**Comment by Araq** [comment]
Yeah, we need some better ``nil`` handling but I mostly wanted to see if anything in "important_packages" would break. ;-)

**Pull request merged**